src/lang/std/engineConnection.ts

```diff
@@ -195,71 +278,243 @@ class EngineConnection {
+                rtc_frames_per_second: 0,
+                rtc_freeze_count: 0,
+                rtc_jitter_sec: 0.0,
+                rtc_keyframes_decoded: 0,
+                rtc_total_freezes_duration_sec: 0.0,
+              }
+
+              // TODO(paultag): Since we can technically have multiple WebRTC
+              // video tracks (even if the Server doesn't at the moment), we
+              // ought to send stats for every video track(?), and add the stream
+              // ID into it.  This raises the cardinality of collected metrics
+              // when/if we do, but for now, just report the one stream.
+
+              videoTrackStats.forEach((videoTrackReport) => {
+                if (videoTrackReport.type === 'inbound-rtp') {
+                  client_metrics.rtc_frames_decoded =
+                    videoTrackReport.framesDecoded || 0
+                  client_metrics.rtc_frames_dropped =
+                    videoTrackReport.framesDropped || 0
+                  client_metrics.rtc_frames_received =
+                    videoTrackReport.framesReceived || 0
+                  client_metrics.rtc_frames_per_second =
+                    videoTrackReport.framesPerSecond || 0
+                  client_metrics.rtc_freeze_count =
+                    videoTrackReport.freezeCount || 0
+                  client_metrics.rtc_jitter_sec = videoTrackReport.jitter || 0.0
+                  client_metrics.rtc_keyframes_decoded =
+                    videoTrackReport.keyFramesDecoded || 0
+                  client_metrics.rtc_total_freezes_duration_sec =
+                    videoTrackReport.totalFreezesDuration || 0
+                } else if (videoTrackReport.type === 'transport') {
+                  // videoTrackReport.bytesReceived,
+                  // videoTrackReport.bytesSent,
+                }
+              })
+              resolve(client_metrics)
+            })
+          })
+        }
+
+        this.mediaStream = mediaStream
+      })
+
+      this.pc.addEventListener('datachannel', (event) => {
+        this.unreliableDataChannel = event.channel
+
+        this.state = {
+          type: EngineConnectionStateType.Connecting,
+          value: {
+            type: ConnectingType.DataChannelConnecting,
+            value: event.channel.label,
+          },
+        }
+
+        this.unreliableDataChannel.addEventListener('open', (event) => {
+          this.state = {
+            type: EngineConnectionStateType.Connecting,
+            value: {
+              type: ConnectingType.DataChannelEstablished,
+            },
+          }
+
+          if (this.shouldTrace()) {
+            dataChannelSpan.resolve?.()
+          }
+
+          this.onDataChannelOpen(this)
+
+          // Do this after we set the connection is ready to avoid errors when
+          // we try to send messages before the connection is ready.
+          this.state = { type: EngineConnectionStateType.ConnectionEstablished }
```

Comment:
A question I asked myself too.

Technically it shouldn't matter, since the video stream is just a view into the engine. The datachannel can start firing off commands immediately and we'll see the results regardless since all state is in the engine and the video track/stream is just a view into it - right?

I think you are right though particularly about the ConnectionEstablished state. To me, telling the rest of the program via `onDataChannel` and other callbacks should be enough to sequence when things happen. I'll look into this further though to be sure

Reply:
It seems there's no obvious way to know when a media stream is live (other than checking MediaStreamTrack readyState) - there's no events that fire when readyState changes :|.

Luckily datachannel is very easy to know when connected. I'll change things around so the track is given after the datachannel is established, and then we can safely set the overall ConnectionEstablished :)

Reply:
@lf94 There is only one event I've found for the MST - which is that `unmute` does fire when the video starts to play, but it felt a bit weird to attach to that event too much. We use `unmute` to clear the loading screen too IIRC

Reply:
Oh god yes, I remember now (after debugging here):

The data channel MUST be opened first. This is because the engine can run in a "headless mode". This is problematic because we get a track connection immediate on `setRemoteDescription` - I think browsers though are bit intelligent and don't try to download any stream data until trying to play the stream, so we don't see issues until we pass off the MediaStream to the other part of the application.

Reply:
Ahh! Yes that makes total sense. Also, they're both part of the same RTCP stream. Yeah, ok. Closing this thread; nice call